core/services/directrequest/delegate.go

```diff
@@ -101,11 +113,76 @@ func (d listener) HandleLog(lb log.Broadcast) {
 		return
 	}
 
-	// TODO: Logic to handle log will go here
+	log := lb.DecodedLog()
+	if log == nil || reflect.ValueOf(log).IsNil() {
+		logger.Error("HandleLog: ignoring nil value")
+		return
+	}
+
+	// TODO: Need to filter each job to the _jobId - can we filter upstream?
+	// TODO: Will need to generate a jobID somehow... hash of DAG?
+	switch log := log.(type) {
+	case *oracle_wrapper.OracleOracleRequest:
+		d.handleOracleRequest(log)
+		err = lb.MarkConsumed()
+		if err != nil {
+			logger.Errorf("Error marking log as consumed: %v", err)
+		}
+	case *oracle_wrapper.OracleCancelOracleRequest:
+		d.handleCancelOracleRequest(log.RequestId)
+		// TODO: Transactional/atomic log consumption would be nice
+		err = lb.MarkConsumed()
+		if err != nil {
+			logger.Errorf("Error marking log as consumed: %v", err)
+		}
+
+	default:
+		logger.Warnf("unexpected log type %T", log)
+	}
+}
+
+func oracleRequestToMap(req *oracle_wrapper.OracleOracleRequest) map[string]interface{} {
+	result := make(map[string]interface{})
+	result["specId"] = fmt.Sprintf("0x%x", req.SpecId)
+	result["requester"] = req.Requester.Hex()
+	result["requestId"] = fmt.Sprintf("0x%x", req.RequestId)
+	result["payment"] = fmt.Sprintf("%v", req.Payment)
+	result["callbackAddr"] = req.CallbackAddr.Hex()
+	result["callbackFunctionId"] = fmt.Sprintf("0x%x", req.CallbackFunctionId)
+	result["cancelExpiration"] = fmt.Sprintf("%v", req.CancelExpiration)
+	result["dataVersion"] = fmt.Sprintf("%v", req.DataVersion)
+	result["data"] = fmt.Sprintf("0x%x", req.Data)
+	return result
+}
+
+func (d *listener) handleOracleRequest(req *oracle_wrapper.OracleOracleRequest) {
+	meta := make(map[string]interface{})
+	meta["oracleRequest"] = oracleRequestToMap(req)
+	ctx := context.TODO()
+	_, err := d.pipelineRunner.CreateRun(ctx, d.jobID, meta)
+	if err != nil {
+		logger.Errorw("DirectRequest failed to create run", "err", err)
+	}
+}
 
-	err = lb.MarkConsumed()
+// Cancels runs that haven't been started yet, with the given request ID
+// TODO: Boy does this ever need testing
+func (d *listener) handleCancelOracleRequest(requestID [32]byte) {
+	err := d.db.Exec(`
+DELETE FROM pipeline_runs
```

Comment:
Maybe push this into the `pipeline.ORM` and add a pass-through method to the `pipeline.Runner`?